### PR TITLE
Schedules queries performance optimization

### PR DESCRIPTION
**Pull request by lboyarsky**
Ticket: [API: investigate slow schedule queries](https://app.asana.com/0/810933294009540/1126143114576745/f)

So basically the problem is that the call from the ticket creates around 150k matches for stop_id + trip_id. So far I don't have better ideas than to create addition index, which would store the {stop_id, trip_id} tuple and use it for querying. This gives around 15%-20% performance improvement (after vs before):

```bash
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
1.215152
1.304598
1.545212
1.282501
1.244547
1.237371
1.226952
1.233228
1.228961
1.260002
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
1.701003
1.449923
1.537593
1.516268
1.550561
1.527346
1.536505
1.509403
1.480983
1.346233
```

Opening this PR as draft (the tests are going to fail) to discuss the implementation.

### CODE REVIEW

**Comment by paulswartz** [comment]
Are those the right URLs? I got 400 errors when I tried them locally.
Other things to look at:
- maybe we can query the route directly, rather than processing it into a list of trips?
- is including the predictions slow, since it needs to happen for each schedule rather than in bulk?

**Comment by lboyarsky** [comment]
I've checked and including prediction doesn't add too much of an overhead (~100ms). I've then created separate handling for routes+stops and performance impact is good:

```bash
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
0.790563
0.737157
0.770308
0.808110
0.807374
0.744893
0.815956
0.791982
0.808808
0.977873
```

I'm, however, not sure if the filtering logic is as expected, all tests passed though.. please take a look.

**Comment by lboyarsky** [comment]
Just realized that it's probably not right as it may include schedules for right stops but for wrong routes.. will continue to look into it.

**Comment by lboyarsky** [comment]
Though now I recall the discussion that stops should be unique to the routes, so may be it still can work?

File: apps/state/lib/state/schedule.ex [hunk 1]
```diff
@@ -141,8 +141,25 @@ defmodule State.Schedule do
     {:ok, overridden_state, timeout_or_hibernate}
   end
 
-  # Converts routes and stops into workable ids
   @spec convert_filters(filter_opts) :: convert_filters
+  defp convert_filters(%{routes: routes, stops: stops} = filters) do
+    route_ids =
+      routes
+      |> Stream.uniq()
+      |> MapSet.new()
```
Comment (paulswartz) [inline apps/state/lib/state/schedule.ex]: `MapSet.new` will already ensure that the values are unique: you don't need the extra check.

File: apps/state/lib/state/schedule.ex [hunk 2]
```diff
@@ -141,8 +141,25 @@ defmodule State.Schedule do
...
+
+    stop_ids =
+      stops
+      |> Stream.uniq()
+      |> Enum.filter(&stop_on_routes?(&1, route_ids))
+
+    filters
+    |> Map.delete(:routes)
+    |> Map.put(:stops, stop_ids)
+    |> convert_filters()
```
Comment (paulswartz) [inline apps/state/lib/state/schedule.ex]: You're right in that this doesn't quite work. Bus stops can have multiple routes which stop there, so this doesn't catch that case. For example, try http://localhost:4000/schedules?filter[route]=9&filter[stop]=44 where you'll also get schedules for the 5 and 10.

**Comment by lboyarsky** [comment]
What about this one? Seems to be working correctly and still gives about 30-40% performance improvements (after vs before):
```bash
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
1.152317
1.020204
1.010872
1.032722
1.002846
0.989446
0.998277
0.984361
0.986858
0.963303
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
1.570320
1.462550
1.457274
1.406484
1.429401
1.413111
1.534585
1.460685
1.416331
1.460762
```

File: apps/state/lib/state/schedule.ex [hunk 3]
```diff
@@ -171,9 +171,23 @@ defmodule State.Schedule do
   defp build_filter_matchers(%{stops: stops, trips: trips} = filters) do
     stop_sequence_matchers = build_stop_sequence_matchers(filters[:stop_sequence])
 
+    all_trips = State.Trip.by_ids(trips)
+    routes_from_trips = all_trips |> Enum.map(& &1.route_id) |> MapSet.new()
```
Comment (paulswartz) [inline apps/state/lib/state/schedule.ex]: `MapSet.new(all_trips, & &1.route_id)`

File: apps/state/lib/state/schedule.ex [hunk 4]
```diff
@@ -171,9 +171,23 @@ defmodule State.Schedule do
   defp build_filter_matchers(%{stops: stops, trips: trips} = filters) do
```
Comment (paulswartz) [inline apps/state/lib/state/schedule.ex]: What about making a matcher for `%{stops: stops, routes: routes}`? Then you wouldn't need to back out from trip to route?

File: apps/state/lib/state/schedule.ex (hunk 3 above)
Comment (lboyarsky) [reply apps/state/lib/state/schedule.ex]: Fixed

File: apps/state/lib/state/schedule.ex (hunk 4 above)
Comment (lboyarsky) [reply apps/state/lib/state/schedule.ex]: I've tried to do it but it works slower.. 

```bash
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
2.219786
2.191719
2.253387
2.123809
2.135842
2.072874
2.101699
2.264730
2.266189
2.193600
```

This is how I did it: 

```elixir
  defp build_filter_matchers(%{stops: stops, routes: routes} = filters) do
    stop_sequence_matchers = build_stop_sequence_matchers(filters[:stop_sequence])

    matchers =
      for stop_id <- stops,
          route_id <- routes,
          stop_sequence_matcher <- stop_sequence_matchers do
        stop_sequence_matcher
        |> Map.put(:route_id, route_id)
        |> Map.put(:stop_id, stop_id)
      end

    %{index: :stop_id, matchers: matchers}
  end
```
Do you have ideas how to implement it better?

Comment (paulswartz) [reply apps/state/lib/state/schedule.ex]: That's the right idea, but also including the part where you use `StopsOnRoute` to ensure that we only build a matcher when the stop is actually on the route.

Comment (lboyarsky) [reply apps/state/lib/state/schedule.ex]: It's still slow (no performance improvements comparing to unchanged version) like that: 

```bash
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
1.510134
1.385606
1.576934
1.653379
1.528449
1.371150
1.360028
1.348777
1.358736
1.346985
```

with 

```elixir
  defp build_filter_matchers(%{stops: stops, routes: routes} = filters) do
    stop_sequence_matchers = build_stop_sequence_matchers(filters[:stop_sequence])

    filtered_routes =
      stops
      |> State.RoutesAtStop.by_stops_and_direction()
      |> MapSet.new()
      |> MapSet.intersection(MapSet.new(routes))

    matchers =
      for stop_id <- stops,
          route_id <- filtered_routes,
          stop_sequence_matcher <- stop_sequence_matchers do
        stop_sequence_matcher
        |> Map.put(:route_id, route_id)
        |> Map.put(:stop_id, stop_id)
      end

    %{index: :stop_id, matchers: matchers}
  end
```

Comment (paulswartz) [reply apps/state/lib/state/schedule.ex]: Can you put `RoutesAtStop` into the loop, something like:
```elixir
  defp build_filter_matchers(%{stops: stops, routes: routes} = filters) do
    stop_sequence_matchers = build_stop_sequence_matchers(filters[:stop_sequence])
    route_set = MapSet.new(routes)

    matchers =
      for stop_id <- stops,
          routes_at_stop = MapSet.new(State.RoutesAtStop.by_stops_and_direction([stop_id]),
          route_id <- MapSet.intersection(routes_at_stop, route_set),
          stop_sequence_matcher <- stop_sequence_matchers do
        stop_sequence_matcher
        |> Map.put(:route_id, route_id)
        |> Map.put(:stop_id, stop_id)
      end

    %{index: :stop_id, matchers: matchers}

Comment (lboyarsky) [reply apps/state/lib/state/schedule.ex]: > Then you wouldn't need to back out from trip to route?

I think we cannot do it as filters always include date ([ref](https://github.com/mbta/api/blob/master/apps/api_web/lib/api_web/controllers/schedule_controller.ex#L128)). It means we still need to figure out the trips for this particular date and then filter them by the routes.

**Comment by paulswartz** [comment]
Is the performance better now that you're preloading the predictions?

**Comment by lboyarsky** [comment]
Yes, it's slightly better, see my comment in other PR - https://github.com/mbta/api/pull/76#issuecomment-506029337

File: apps/state/lib/state/schedule.ex [hunk 5]
```diff
@@ -176,9 +176,23 @@ defmodule State.Schedule do
...
+
+    filtered_routes =
+      stops
+      |> State.RoutesAtStop.by_stops_and_direction()
+      |> MapSet.new()
+      |> MapSet.intersection(routes_from_trips)
+
+    filtered_trips =
+      all_trips
+      |> Stream.filter(&Enum.member?(filtered_routes, &1.route_id))
```
Comment (paulswartz) [inline apps/state/lib/state/schedule.ex]: You can use `MapSet.member?/2` here since you know `filtered_routes` is a `MapSet`.

Comment (lboyarsky) [reply apps/state/lib/state/schedule.ex]: Fixed.

**Pull request merged**